### PR TITLE
Plugins: Display plugin permissions required

### LINKED ISSUE #5990: Annotation fetch system enhancement, enable the possibility of per panel annotations
Refactor AnnotationsSrv to make it possible for panels to have specific annotations.

### PR DESCRIPTION
**Pull request by linoman**
**What is this feature?**

This PR adds a new `IAM` tab to the plugin's details page. 
It shows the required permissions (scope and actions) needed by a plugin to interact with Grafana.

Additional info might be added to this tab. Mainly which Service account / token is tied to the plugin (only available after installation)

What's missing?
- [x] Permissions used after installing the plugin
- [x] Rename the property to use the same name as the backend
- [x] ~Update grafanacom to return the iam information~ No need as we have the plugin json with `useFetchDetails`

To test this, you can either install the app-with-service-account plugin example locally.

Or @linoman did a commit with fake data:
*IMPORTANT NOTE: IN ORDER TO TEST THIS FRONTEND WORK, A COMMIT WITH FAKE DATA HAS BEEN ADDED. DROP THE FOLLOWING COMMIT BEFORE MERGING.*

```
3bd71b7538b24a4ee2ca218af200d7cf78931421
```

Fixes grafana/grafana-enterprise#5990 

Preview
image

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by vesnadean** [comment]
Looks good. I have a nitpicky UI comment: we could use more padding between the sentence "The Alert list plugin..." and the first table row.

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (gamab) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:55]: @Clarity-89 I'm using `VerticalGroup` and `HorizontalGroup` here, is it fine? Should I remove the outer `div`?

Comment (Clarity-89) [reply public/app/features/plugins/admin/components/PluginDetailsBody.tsx:55]: We now have Stack component, which we recommend using instead of the Vertical and Horizontal groups. `<Stack direction='column'/>` is equivalent to `VerticalGroup`. Also yes, container shouldn't be necessary here.

**Review by andresmgot — COMMENTED** [review]
It LGTM but I'd wait for a review from someone from @grafana/plugins-platform-frontend just in case

**Comment by andresmgot** [comment]
> Update grafanacom to return the iam information

If you request a version (e.g. https://grafana.com/api/plugins/grafana-athena-datasource/versions/2.13.3) you get the JSON info of the plugin with the IAM information if it's there.

**Comment by gamab** [comment]
> If you request a plugin version (e.g. https://grafana.com/api/plugins/grafana-athena-datasource/versions/2.13.3) you get the JSON info of the plugin with the IAM information if it's there.

Thanks for pointing this out, that made me realize that in the `RemotePlugin` type the plugin `json` was there already!

File: public/app/features/plugins/admin/helpers.ts
Comment (andresmgot) [reply public/app/features/plugins/admin/helpers.ts:222]: probably using the wrong model here, as far as I can see `remote` is populated from a call to https://grafana.com/api/plugins which doesn't include JSON info.

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (Clarity-89) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:58]: Would the[ InteractiveTable component ](https://grafana-dev.com/developers/saga/templates/table) work here?

File: public/app/features/plugins/admin/helpers.ts
Comment (gamab) [reply public/app/features/plugins/admin/helpers.ts:222]: Yes, as I was telling @linoman we'll need to piggy bag on the `useFetchDetails` hook that does the query to https://grafana.com/api/plugins/PLUGIN_ID/ when you navigate to the plugin details page.

In a first iteration: If the plugin is installed we'll get the permissions through `plugin.iam` (populated when we get the plugin list from the instance), if the plugin is not installed we'll get the permissions through `plugin.json.iam`.

Comment (linoman) [reply public/app/features/plugins/admin/helpers.ts]: Changed as requested

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (gamab) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:66]: I think it can be `plugin.iam?.permissions` from local plugin or `plugin.details?.iam?.permissions` for a fetched plugin?

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (linoman) [reply public/app/features/plugins/admin/components/PluginDetailsBody.tsx]: Changed as requested! Thanks for the help @Clarity-89

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (Clarity-89) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx]: Should this feature toggle and the page condition still be used?

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (Clarity-89) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:22]: This type can be defined outside the component.

**Review by Clarity-89 — APPROVED** [review]
Looks good, left two minor comments.

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (linoman) [reply public/app/features/plugins/admin/components/PluginDetailsBody.tsx]: Changed as requested.

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (linoman) [reply public/app/features/plugins/admin/components/PluginDetailsBody.tsx]: Changed as requested.

**Review by gamab — APPROVED** [review]
LGTM! Thanks Lino :)

**Comment by gamab** [comment]
@grafana/plugins-platform-frontend Can we get your eyes on this one? :pray: :slightly_smiling_face:

**Review by sunker — APPROVED** [review]
LGTM

File: public/app/features/plugins/admin/api.ts
Comment (jackw) [inline public/app/features/plugins/admin/api.ts:29]: nit: doesn't [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) mean this will result in `undefined` if remote or json aren't defined? Not sure what the need is for the nullish coalesce. 🤔

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (jackw) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:79]: nit: can we put this logic in a variable so it's easier to scan the code?

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (jackw) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:34]: Shouldn't this be `cell: ({ cell: { value } }: Cell<'action'>) => value,`?

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (jackw) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:39]: Should this be `cell: ({ cell: { value } }: Cell<'scope'>) => value,`?

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (jackw) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:28]: What's the need for the expect-error?

Comment (academo) [inline public/app/features/plugins/admin/components/PluginDetailsBody.tsx:28]: Could you work this type instead of ignore the typescript error?

**Comment by academo** [comment]
I have a general question regarding the accuracy of the tab itself. Are we maybe only displaying "backend" permissions only? 

The frontend part of a pluggn can do almost any operation in behalf of the user by calling the necessary APIs from the browser so this tab might create a false sense of security. e.g.: "This plugin doesn't have `dashboard:read` so my queries are safe" but in reality, they are only safe in the backend part of the plugin?

I understand there's a big limitation on how you can control what the frontend part can do in behalf of the user so this section needs to clarify what this permissions are and how they work better. E.g. link to more information of what these scopes mean, make sure the user understand that they are not as fail-proof as "no scope = no access at all times"

File: public/app/features/plugins/admin/helpers.ts
Comment (jackw) [inline public/app/features/plugins/admin/helpers.ts:222]: This function munges the local installed plugin data (that comes from backend) with the remote plugin data (that comes from gcom api). Here we have the opportunity to fall back from local to remote when a plugin is or isn't installed. If my memory serves me correctly this, currently, will prevent rendering iam information for any uninstalled plugins. Is that intended?

**Review by jackw — COMMENTED** [review]
Awesome work on this @linoman . I've left some comments / nits.

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (academo) [reply public/app/features/plugins/admin/components/PluginDetailsBody.tsx:34]: Indeed. And this also fixes the typescript problem that is ignored in this variable declaration.

Comment (Clarity-89) [reply public/app/features/plugins/admin/components/PluginDetailsBody.tsx:34]: Damn, this is probably it. Spent a while yesterday trying to figure out the cause of that type error 😅

Comment (linoman) [reply public/app/features/plugins/admin/components/PluginDetailsBody.tsx]: Great eye @jackw 

We spent some time trying to figure this one out and because of the data structures needed for `InteractiveTable` we missed this. This also helps with the ugly `//ts-expect-error`

Thank you for your review!

File: public/app/features/plugins/admin/components/PluginDetailsBody.tsx
Comment (linoman) [reply public/app/features/plugins/admin/components/PluginDetailsBody.tsx]: Changed as requested

File: public/app/features/plugins/admin/api.ts
Comment (linoman) [reply public/app/features/plugins/admin/api.ts:29]: If both results are undefined, that means the plug-on doesn't require any permissions. Do you suggest an empty object instead?

File: public/app/features/plugins/admin/helpers.ts
Comment (linoman) [reply public/app/features/plugins/admin/helpers.ts]: Changed as requested.

File: public/app/features/plugins/admin/api.ts
Comment (jackw) [reply public/app/features/plugins/admin/api.ts:29]: What I mean is `remote?.json?.iam` should return `undefined` unless the properties exist. Which is why I'm wondering if we need the `?? undefined` part. There's no harm in it I'm just being curious.

image

>The optional chaining (?.) operator accesses an object's property or calls a function. If the object accessed or function called using this operator is [undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined) or [null](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/null), the expression short circuits and evaluates to [undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined) instead of throwing an error.

File: public/app/features/plugins/admin/helpers.ts
Comment (linoman) [reply public/app/features/plugins/admin/helpers.ts]: I shouldn't have changed this. We had a conversation about this [[link](https://github.com/grafana/grafana/pull/78355/commits/7dcffb833186f67f85370bf1b6a5c517c96efb52#r1422420663)]

This has been reverted

**Comment by gamab** [comment]
@academo Thanks for sharing your thoughts :+1: 

> I have a general question regarding the accuracy of the tab itself. Are we maybe only displaying "backend" permissions?

Yes that's correct, we only display backend permissions.

> The frontend part of a plugin can do almost any operation in behalf of the user by calling the necessary APIs from the browser so this tab might create a false sense of security. e.g.: "This plugin doesn't have dashboard:read so my queries are safe" but in reality, they are only safe in the backend part of the plugin?
I understand there's a big limitation on how you can control what the frontend part can do in behalf of the user so this section needs to clarify what these permissions are and how they work better.

Fair point, we're changing the message a bit to state that the permissions are related to the plugin **backend**. However I'm not sure we should explicitly point out that the frontend can still make any request it wants, should we?

> E.g. link to more information of what these scopes mean, make sure the user understand that they are not as fail-proof as "no scope = no access at all times"

I'm not sure what you mean by this. Do you want us to link to the Grafana RBAC documentation?

**Review by academo — APPROVED** [review]
I had a brief discussion with @linoman regarding clarifying what these permissions means for a plugin. 

Generally speaking clarifying that these permissions only affect the behaviour of the plugin backend running unattended in the grafana instance (as with the current latest message) and adding a link to an explanation of how RBAC affects plugins will help.

@gamab I completely agree pointing out that the frontend can do almost anything in behalf of the user is not necessary.

File: public/app/features/plugins/admin/api.ts
Comment (linoman) [reply public/app/features/plugins/admin/api.ts]: You are right. If either values of the optional chaining are not defined, the end result will be undefined. I have made the changes as suggested.

**Comment by linoman** [comment]
@academo As we discussed, I updated the message and created an [issue](https://github.com/grafana/grafana-enterprise/issues/6143) for linking the new IAM section with the docs

**Pull request merged**
